src/simulation_strategies/GeneralizedCliffordsSimulationStrategy.py

```diff
@@ -0,0 +1,164 @@
+__author__ = 'Tomasz Rybotycki'
+
+from src.simulation_strategies.SimulationStrategy import SimulationStrategy
+from numpy import ndarray, zeros, array
+from numpy.random import choice
+from numpy.linalg import norm
+from math import factorial
+from typing import List
+from copy import copy
+from src.Boson_Sampling_Utilities import calculate_permanent
+
+
+class GeneralizedCliffordsSimulationStrategy(SimulationStrategy):
+    def __init__(self, interferometer_matrix: ndarray):
+        self.r_sample = []
+        self.number_of_input_photons = 0
+        self.pmfs = []  # Probability mass functions calculated along the way.
+        self.interferometer_matrix = interferometer_matrix
+        self.input_state = []
+        self._labeled_states = dict()
+        self.current_outputs = []
+
+    def simulate(self, input_state: ndarray) -> List[int]:
+        """
+            Returns sample from linear optics experiments given output state.
+            :param input_state: Input state in particle basis.
+            :return: A resultant state after traversing through interferometer.
+        """
+        self.input_state = input_state
+        self.number_of_input_photons = sum(input_state)
+        self.__get_sorted_possible_states()
+        self.__fill_r_sample()
+        return self.r_sample
+
+    def __get_sorted_possible_states(self) -> None:
+        """
+            Calculate and sort all the substates of the input. They will later be used to calculate output probabilities.
+            :return: Dict of all the possible substates of input (including 0 and the state), labeled with number of
+            particles in this state.
+        """
+        # Calculating all possible substates of the input
+        possible_input_states = self.__calculate_all_input_substates(self.input_state[:])
+
+        # Labeling them into dict where keys are being number of particles in the state.
+        self._labeled_states = dict()
```

Comment:
Consideration: https://docs.python.org/2/library/collections.html#collections.defaultdict – maybe you'll find this interesting.

Reply:
I've changed it. That's quite clever.